Insert URL Modal

aliseyalvi opened the pull request:
### What does this PR?
fixes #2208 .
- added modal for inserting url while writing post in editor
- takes label and url from input and formats it.
- insert raw url if no label is provided


### Where should the reviewer start?

- Go to editor screen
- tap on link button in editor footer.
- modal opens with rspective fields.

### Issue number
#2208 

### Screenshots/Video

https://user-images.githubusercontent.com/48380998/157290599-1790d053-5e45-4022-8fb8-41a613510f0f.mp4

noumantahir reviewed (commented):
So far so good, I do believe using apply Snippet to insert the formatted text and formatting text inside apply link modal with simplify things for us a lot considering we will be using formatted texts for preview and also in quick comment modal moving forward... Let me know your thoughts on this.

Also,

- when opening link modal, try to keep the keyboard open 
- change `keyboardType` of url input field to `url`
- after insertion, the cursor should move towards the end of inserted string. applySnippet should resolve this by default though.

aliseyalvi commented:
@noumantahir check the latest update here.
- added support for plain, video and image type urls
- added support for preview of video and images
- added select text/url to prefill the label/url and insert the other.
- minor bugs fixed as well.
- 
https://user-images.githubusercontent.com/48380998/159011370-225dadf0-d1d7-4738-bae3-8cc2b084d79b.mp4

noumantahir commented on src/components/markdownEditor/view/markdownEditorView.js:
```diff
@@ -290,6 +296,31 @@ const MarkdownEditorView = ({
+  };
+  const _handleInsertLink = ({ label, url, selection, isImage }) => {
+    if (url) {
+      applyWebLinkFormat({
```
I still strongly thing the `applySnippet` will serve better in this case, can we can simply cook the link text within `insertLinkModal` that would passed just the text to be added in body... let me know if you need to discuss this in details.. it would also help you render plain link preview inside insertLinkModal

aliseyalvi replied:
actually the applySnippet doesn't format the url text according to our need, like we need [label](url) this format for labeled urls and ![name][(img link) this format for images. And for that purpose we need applyWebLinkFormat and I had used that to accomplish the desired goal.

noumantahir reviewed (commented):
@aliseyalvi Check code related comments above....

@feruzm  On the UI end, how about we keep the preview window always visible, also have a text preview of markdown that will be added to editor. For instance...

Label Input - input
URL - input
Markdown - preview markdown text
Preview - html rendering

noumantahir commented:
@aliseyalvi take it easy, always test each feature you implement aggressive with an intention to break you own code....

![Screenshot 2022-03-20 at 7 57 58 PM](https://user-images.githubusercontent.com/6298342/159169824-94cbaff7-dad4-49f6-bbb5-a10e39e23332.png)
![Screenshot 2022-03-20 at 7 59 04 PM](https://user-images.githubusercontent.com/6298342/159169826-65482027-1773-433c-a3a2-ea713338ba6b.png)
![Screenshot 2022-03-20 at 8 00 30 PM](https://user-images.githubusercontent.com/6298342/159169829-fc277c46-83cf-4fe8-9df2-85e65ddcd46a.png)
![Screenshot 2022-03-20 at 8 02 54 PM](https://user-images.githubusercontent.com/6298342/159169831-517f58df-ff8a-4d1c-821b-340c2729e88b.png)
![Screenshot 2022-03-20 at 8 04 21 PM](https://user-images.githubusercontent.com/6298342/159169834-fff09856-3565-4598-8989-286b02a24aca.png)
![Screenshot 2022-03-20 at 8 08 47 PM](https://user-images.githubusercontent.com/6298342/159169835-766135e1-6494-4e5b-8a14-86316da0cc92.png)
![Screenshot 2022-03-20 at 8 10 33 PM](https://user-images.githubusercontent.com/6298342/159169837-df31d56d-9774-46fe-b28e-1785399f8cdf.png)
![Screenshot 2022-03-20 at 8 18 35 PM](https://user-images.githubusercontent.com/6298342/159169840-83d62be4-285b-47d3-9d24-4e4bf0c10e61.png)
![Screenshot 2022-03-20 at 8 21 43 PM](https://user-images.githubusercontent.com/6298342/159169841-b5073eb5-54e3-4934-aa73-9dfed94c94b1.png)
![Screenshot 2022-03-20 at 8 22 30 PM](https://user-images.githubusercontent.com/6298342/159169842-c4e7d3fb-6fd3-4754-a643-dca3a4a3e0fa.png)
![Screenshot 2022-03-20 at 8 24 25 PM](https://user-images.githubusercontent.com/6298342/159169843-1db45bb9-6ef5-4805-804a-adec62c14abf.png)

aliseyalvi commented:
@noumantahir PR updated.

- fixed preview not changing issue
- fixed label not updating issue
- fixed url not showing complete issue
- removed url label from plain url
- added line breaks before and after image and video
- added side effect for each change e-g label/url/urltype and update the preview

noumantahir commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+            iconName="plus"
+            iconType="MaterialCommunityIcons"
+            iconColor="white"
+            text={'Insert Link'}
```
use intl

aliseyalvi replied:
https://github.com/ecency/ecency-mobile/pull/2216/commits/891fb1258ad83f0423289d71e2f6975b118eccf1

noumantahir commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+            <Text style={styles.previewText}>Preview</Text>
+            {previewBody ? (
+              <View style={styles.preview}>
+                <PostBody body={previewBody} onLoadEnd={() => setIsLoading(false)} />
```
links are still tappable... ignore taps inside.

aliseyalvi replied:
https://github.com/ecency/ecency-mobile/pull/2216/commits/7baffe9f3e0c9ea17f019b220000f4005ed783b0

noumantahir reviewed (changes requested):
@aliseyalvi greate work, just some small touch ups before we can finalise this.

aliseyalvi commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+          <View style={styles.previewContainer}>
+            <Text style={styles.previewText}>Preview</Text>
+            {previewBody ? (
+              <View style={styles.preview}>
```
https://github.com/ecency/ecency-mobile/pull/2216/commits/4fdc75e613f0c32d43661ae93eceab29a12e9e85

aliseyalvi commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+          <TextButton
+            style={styles.cancelButton}
+            onPress={() => sheetModalRef.current?.setModalVisible(false)}
+            text={'Cancel'}
```
https://github.com/ecency/ecency-mobile/pull/2216/commits/891fb1258ad83f0423289d71e2f6975b118eccf1

aliseyalvi commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+          style={styles.input}
+          value={url}
+          onChangeText={_handleUrlChange}
+          placeholder={'Enter URL'}
```
https://github.com/ecency/ecency-mobile/pull/2216/commits/891fb1258ad83f0423289d71e2f6975b118eccf1

aliseyalvi commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+          style={styles.input}
+          value={label}
+          onChangeText={_handleLabelChange}
+          placeholder={'Enter Label (Optional)'}
```
https://github.com/ecency/ecency-mobile/pull/2216/commits/891fb1258ad83f0423289d71e2f6975b118eccf1

aliseyalvi commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+const URL_TYPES = [
+  {
+    id: 0,
+    title: 'plain',
```
https://github.com/ecency/ecency-mobile/pull/2216/commits/67a2040d8be27be6358cedf921396822061c3be8

noumantahir commented on src/components/insertLinkModal/insertLinkModal.tsx:
```diff
@@ -0,0 +1,242 @@
+      }
+      if (url) {
+        const labelText =
+          selectedUrlType === 2 ? 'dummy-label' : selectedUrlType === 1 ? '' : label; //TODO: replace image label here. Using demo label for now
```
Just extract last endpoint /path from url

Example: https://noumantahir.com/image.png
Label = image.png


Check image selection modal, that might have something similar

aliseyalvi commented:
@noumantahir PR updated.

- added clipboard listener. When insert model opens it checks for clipboard and if there is a valid urls it prefills the url input.
- fixed preview View height with scrollview support to contain large and portrait images/videos
- added image label when image url is inserted.

noumantahir reviewed (approved):
seems to work acceptable to this point.

@feruzm looking forward to your feedback, there is a small thing that user needs to hide keyboard to fully see preview on image/video on android, but I would mark that as acceptable thing keeping in view it's taking too much time refining such a small behaviour that may not mean much to most of bunch

feruzm reviewed (approved):
lgtm, let's do some tests on device.
